### PR TITLE
sql: add extra information to protocol errors in bind

### PR DESCRIPTION
**Pull request by cucaroach**
A user is running into mysterious protocol errors when using prepared
statements. Add some extra information to the error message to help
guide the investigation.

Informs: https://github.com/cockroachlabs/support/issues/2184
Release note: none
Epic: none

### CODE REVIEW

File: pkg/sql/conn_executor_prepare.go
Comment (knz) [inline pkg/sql/conn_executor_prepare.go:349]: nit: pls consider `errors.WithDetail` instead of copying the full SQL syntax as a prefix.

File: pkg/sql/conn_executor_prepare.go
Comment (knz) [inline pkg/sql/conn_executor_prepare.go:364]: ditto

File: pkg/sql/conn_executor_prepare.go
Comment (cucaroach) [reply pkg/sql/conn_executor_prepare.go]: 👍

File: pkg/sql/conn_executor_prepare.go
Comment (cucaroach) [reply pkg/sql/conn_executor_prepare.go]: 👍

**Comment by cucaroach** [comment]
@rafiss @knz what I can't figure out is how I get this error message into the logs.   Nothing I've tried gets this error message into the crdb logs.  I'm running a PG test with the server running and I tried sql.trace.session_eventlog.enabled, server.auth_log.sql_sessions.enabled, server.auth_log.sql_connections.enabled, sql.trace.log_statement_execute and nothing seems to work.  Any help/ideas would be most welcome!  

Running test like this:

```
./dev test pkg/sql/pgwire -f TestPGTest/bind_proto_error --test-args "-addr localhost:26257 -user root" -v
```

Where bind_proto_error just contains the new test I added to errors to hit this error path. Ideally we'd just roll this patch into an upgrade and error message would show up in logs the next time it gets hit.

**Review by rafiss — COMMENTED** [review]
The `Detail` field of an `ErrorResponse` is separate from the `Message` field. See: https://github.com/cockroachdb/cockroach/blob/d04dde517c69ac2216777b7e1b8476a5eae93572/pkg/testutils/pgtest/pgtest.go#L199-L217

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106130#-:-NZg1Diq0W9mOiBq4d7p:bc6kudz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @rytaft)

___
*[`pkg/sql/conn_executor_prepare.go` line 349 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106130#gh-1253622286:-NZg2MRw5qvnDLdjvaHg:b5o0nx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6930cf1dadb5c007d078149fa0f609a8faba3f8/pkg/sql/conn_executor_prepare.go#L349)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

👍 
</blockquote></details>

nit: the word "failed" is probably redundant in the detail message
___
*[`pkg/sql/conn_executor_prepare.go` line 364 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106130#gh-1253622368:-NZg0iEIDffldq6ec2_7:bagga) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6930cf1dadb5c007d078149fa0f609a8faba3f8/pkg/sql/conn_executor_prepare.go#L364)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

👍 
</blockquote></details>

nit: i would say `statement summary` instead of `statement`, so that this message is distinguished from the one above
___
*[`pkg/sql/conn_executor_prepare.go` line 362 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106130#-NZg2WxI-PrLz-SKpWvX:-NZg2WxI-PrLz-SKpWvY:b5677v2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca0ae3976f10f5835a9951c175f4a181b5c71b6d/pkg/sql/conn_executor_prepare.go#L362)):*
> ```Go
> 	}
> 
> 	retErr := func(err error) (fsm.Event, fsm.EventPayload) {
> ```

nit: i'd prefer not to have two `retErr` functions. could we just add this to the first one above:
```
		if ps != nil && ps.StatementSummary != "" {
			err = errors.WithDetailf(err, "statement (%q) failed", ps.StatementSummary)
		}
```

(and add a `var ps *PreparedStatement` declaration above)

**Comment by cucaroach** [comment]
This is RFAL, I added a log statement where the mysterious error is occurring, these errors are logged by default and I wanted to be able to search for these errors in logs. Maybe there's a better way?

**Comment by yuzefovich** [comment]
We got the same support ticket again over in https://github.com/cockroachlabs/support/issues/2510 - should we push this over the finish line and backport? @cucaroach @rafiss @knz

**Comment by cucaroach** [comment]
I think so.  I think I addressed all the review feedback so just waiting on review approval.

**Review by rafiss — APPROVED** [review]
> Reviewed 2 of 7 files at r1, 4 of 5 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106130)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @rytaft)

**Review by rafiss — APPROVED** [review]
yeah, i think we should try to get this out. thanks for your improvements @cucaroach, and @yuzefovich for following up

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106130#-:-NbLY7PKESqFZTIzg6kV:b2gt3pz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @rytaft)

**Comment by rafiss** [comment]
general tip: if you click the "re-request review" button, then Github reminders would kick in to prod me to review this so you don't have to :)

image

**Comment by cucaroach** [comment]
> general tip: if you click the "re-request review" button, then Github reminders would kick in to prod me to review this so you don't have to :)

TIL!   Thanks!

**Comment by cucaroach** [comment]
bors r=rafiss

**Pull request merged**
